### PR TITLE
Date utility commands

### PR DESCRIPTION
**Pull request by jz448**
For #2324. Based on the discussions in the linked issue, I have made the following modifications on `date` related commands:

 - changed the primitive `Date(DateTime<Utc>)` type to `Date(DateTime<FixedOffset>)` to handle different time zone requirements
 - `date now` now passes on the `Date` type
 - `date format` now prints formatted string directly by default and gives an useful Nushell error message when the format string is invalid instead of panicking
 - added `date to-table` command to print a date in a structured table
 - added `date to-timezone` now supports several common time zone abbreviations, including UTC and LOCAL, and also format strings such as `+0600`, `-04:30`, `+02` 
 - removed `date utc` command since it's equivalent to `date now | date to-timezone UTC`

### CODE REVIEW

File: crates/nu-cli/src/commands/date/format.rs [hunk 1]
```diff
@@ -24,11 +24,11 @@ impl WholeStreamCommand for Date {
     fn signature(&self) -> Signature {
         Signature::build("date format")
             .required("format", SyntaxShape::String, "strftime format")
-            .switch("raw", "print date without tables", Some('r'))
+            .switch("table", "print date in a table", Some('t'))
```
Comment (jonathandturner) [inline crates/nu-cli/src/commands/date/format.rs:27]: What does the --table switch do?

File: crates/nu-cli/src/commands/date/parser.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    if s.chars().all(|x| x.is_ascii_alphabetic()) {
+        match s.to_lowercase().as_str() {
+            "gmt" | "utc" | "ut" => offset_hours(0),
+            "edt" => offset_hours(-4),
+            "est" | "cdt" => offset_hours(-5),
+            "cst" | "mdt" => offset_hours(-6),
+            "mst" | "pdt" => offset_hours(-7),
+            "pst" => offset_hours(-8),
+            "local" => Ok(*Local::now().offset()),
```
Comment (jonathandturner) [inline crates/nu-cli/src/commands/date/parser.rs:31]: Is there a date function or crate we can use to do this? Only allowing utc and US timezones seems limited

File: crates/nu-cli/src/commands/date/to_timezone.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            "time zone",
+            SyntaxShape::String,
+            "time zone description",
+        )
+    }
+
+    fn usage(&self) -> &str {
+        "Convert a date to a given time zone.
+        
+Supported time zone abbreviations include: GMT, UTC, UT, EDT, EST, CDT, CST, MDT, MST, PDT, PST, LOCAL.
```
Comment (jonathandturner) [inline crates/nu-cli/src/commands/date/to_timezone.rs:32]: This is another one we might be able to use a crate or built-in date function to help, so that we can support all the timezones.

File: crates/nu-cli/src/types/deduction.rs [hunk 4]
```diff
@@ -765,60 +765,59 @@ impl VarSyntaxShapeDeductor {
...
                     )?;
                 }
-                Operator::In | Operator::NotIn => {
-                    match var_side {
-                        BinarySide::Left => match &expr.expr {
-                            Expression::List(list) => {
-                                if !list.is_empty() {
-                                    let shapes_in_list = self
-                                        .get_shapes_in_list_or_insert_dependency(
+                Operator::In | Operator::NotIn => match var_side {
```
Comment (jonathandturner) [inline crates/nu-cli/src/types/deduction.rs:768]: Why is the deduction code changing? If you need to update it, may make sense to put it into a separate PR and just have this one focused on the date changes.

File: crates/nu-data/src/base/shape.rs [hunk 5]
```diff
@@ -31,7 +31,7 @@ pub enum InlineShape {
     ColumnPath(ColumnPath),
     Pattern(String),
     Boolean(bool),
-    Date(DateTime<Utc>),
+    Date(DateTime<FixedOffset>),
```
Comment (jonathandturner) [inline crates/nu-data/src/base/shape.rs:34]: Can we also do this change as a separate PR? Changing this should probably also change how dates are handled in other commands, like `date now`.

File: crates/nu-cli/src/commands/date/format.rs (hunk 1 above)
Comment (jz448) [reply crates/nu-cli/src/commands/date/format.rs:27]: The original `date format` outputs the result in a table by default. I find it bit inconsistent with the new `date now` and `date to-timezone` commands, which print a string now, so I moved it to a switch instead. I personally don't think the switch is very helpful and happy to remove it.

File: crates/nu-cli/src/commands/date/parser.rs (hunk 2 above)
Comment (jz448) [reply crates/nu-cli/src/commands/date/parser.rs:31]: We could directly use the [parse_from_rfc2822](https://docs.rs/chrono/0.4.19/chrono/struct.DateTime.html#method.parse_from_rfc2822) API in the chrono crate to parse an RFC 2822 timezone. However, this API requires a complete date string input (e.g., "Wed, 18 Feb 2015 23:16:09 GMT") and  therefore it has the overhead of parsing other part that we don't care about (e.g., "Wed, 18 Feb 2015 23:16:09"). Also, the `parse_from_rfc2822` API doesn't support omitting the colon or the minutes part, nor does it support the `UTC` and `LOCAL` timezone, so additional operations need to be added on top of it.

File: crates/nu-cli/src/commands/date/to_timezone.rs (hunk 3 above)
Comment (jz448) [reply crates/nu-cli/src/commands/date/to_timezone.rs:32]: I don't think it's possible to support all time zone abbreviations since the same abbreviation can mean different time zones (also discussed in [this issue](https://github.com/chronotope/chrono-tz/issues/44)). For example, `CST` could mean "Central Standard Time" and "China Standard Time". Therefore, I've only included obsolete time zones supported in [RFC 2822](https://tools.ietf.org/html/rfc2822) and added `UTC` and `LOCAL` because they were mentioned in the discussions. ISO 8601/RFC 3339 doesn't support abbreviations at all.

File: crates/nu-cli/src/types/deduction.rs (hunk 4 above)
Comment (jz448) [reply crates/nu-cli/src/types/deduction.rs:768]: This is changed by `cargo fmt --all`. When I submitted this PR, the Azure Pipelines checkes reported code formatting issues issues so I run that command to fix it. I didn't do any manual change on it. Should I submit a separate PR for this?

File: crates/nu-data/src/base/shape.rs (hunk 5 above)
Comment (jz448) [reply crates/nu-data/src/base/shape.rs:34]: We could keep `Date(DateTime<Utc>)` in this PR but I'm not quite sure what is the benefit? My understanding is that the `InlineShape` enum in `shape.rs` should match the `Primitive` enum, or there will be unnecessary data type conversions between `Date<DateTime<Utc>>` and `Date(DateTime<FixedOffset>)`. If I do the change in a separate PR, then I need to do these conversions in `shape.rs` in this PR.

Also, all date related subcommands including `date now` are now using the `Date(DateTime<FixedOffset>)` type in this PR, so I don't see an issue there. Maybe I missed something? What do you think?

File: crates/nu-cli/src/types/deduction.rs (hunk 4 above)
Comment (jonathandturner) [reply crates/nu-cli/src/types/deduction.rs:768]: That's interesting. Are you using the nightly compiler or the latest stable? Generally we use stable, but I've noticed that sometimes they do release formatting changes.

Comment (jz448) [reply crates/nu-cli/src/types/deduction.rs:768]: I'm running `cargo 1.48.0 (65cbdd2dc 2020-10-14)`. I did some tests earlier: 

- check out the commit `e299e76fcf7a17dabc640a0ff6c02f53fa6548c6` ([commit link](https://github.com/nushell/nushell/commit/e299e76fcf7a17dabc640a0ff6c02f53fa6548c6)) , which is the last commit before my fork of the project
- run `cargo fmt --all -- --check`
- got the formatting issue report

**The current repo no longer have this problem because it has been fixed in a [recent commit](https://github.com/nushell/nushell/commit/b193303aa34fb648dd34068131de28fa7cc98fd9#diff-b6131e0fd615bf0e97bd318197f154553c07306adbe2dac3ef52750448eab430).**

File: crates/nu-cli/src/commands/date/format.rs (hunk 1 above)
Comment (fdncred) [reply crates/nu-cli/src/commands/date/format.rs:27]: So, this reverses the functionality of `raw`? i.e. by default date prints strings and if you want it in a table you pass the `table` parameter. Do I understand this correctly?

File: crates/nu-cli/src/commands/date/parser.rs (hunk 2 above)
Comment (fdncred) [reply crates/nu-cli/src/commands/date/parser.rs:31]: Is [parse_from_rfc3339](https://docs.rs/chrono/0.4.19/chrono/struct.DateTime.html#method.parse_from_rfc3339) an option? It seems more intuitive than 2822 to me. Maybe just put another `else if` and check the length of string or the right number of dashes or some minimal validation that it is a 3339 string.

**Comment by fdncred** [comment]
First of all, let me say, Thank you! Lots of interesting changes here. I've added a few comments.

File: crates/nu-cli/src/commands/date/parser.rs (hunk 2 above)
Comment (jz448) [reply crates/nu-cli/src/commands/date/parser.rs:31]: @fdncred I do agree that RFC3339 looks more intuitive, but here we are only interested in converting an user-provided timezone to an offset that we can use. For example, with the `date to-timezone PST` or `date to-timezone -08:00` or `date to-timezone -0800` or `date to-timezone -08` command, we would like to know that the expected timezone is UTC-08:00. 

If we use either the [parse_from_rfc2822](https://docs.rs/chrono/0.4.19/chrono/struct.DateTime.html#method.parse_from_rfc2822) or the [parse_from_rfc3339](https://docs.rs/chrono/0.4.19/chrono/struct.DateTime.html#method.parse_from_rfc3339) API for this purpose, we simply prefix a fixed date time string (e.g., "Wed, 18 Feb 2015 23:16:09 " for RFC2822 or "1996-12-19T16:39:57" for RFC3339) to the user-provided timezone string, invoke the API, and then extract the parsed timezone offset. In this sense, the format of the fixed date time string doesn't really matter. Nevertheless, using these APIs incur the cost of parsing the fixed date time string that we don't care about. That's why I extracted the code responsible for parsing the timezone from the `chrono` crate (which includes private functions) and made some modifications to suit our use case (e.g., support `date to-timezone LOCAL`).

The reason I suggested the `parse_from_rfc2822` API is because it supports a few timezone abbreviations while RFC3339 supports none. Based on the discussions in #2324, it seems we should support commands like `date to-timezone UTC` and `date to-timezone PST` so timezone abbreviations (well, some of them) should be supported.

**Comment by jz448** [comment]
> First of all, let me say, Thank you! Lots of interesting changes here. I've also added a few comments.

Thanks! I hope my responses to your comments make sense. It's a cool project! Please let me know if you have more comments :D

File: crates/nu-cli/src/commands/date/parser.rs (hunk 2 above)
Comment (fdncred) [reply crates/nu-cli/src/commands/date/parser.rs:31]: Thanks for explaining it to me. I like the idea of using your example, `date to-timezone PST` or `date to-timezone -08:00` or `date to-timezone -0800` or `date to-timezone -08`. It seems like a good compromise to me and it's inclusive of any timezone I think. It appears that your code already works this way, right? If so, I vote we commit this.

File: crates/nu-cli/src/commands/date/to_timezone.rs (hunk 3 above)
Comment (fdncred) [reply crates/nu-cli/src/commands/date/to_timezone.rs:32]: @jz448 - I see two resolutions to this. Please let me know what you think.
1. Doing a bit more research to see if there's a better way to handle timezones that somehow disambiguates the abbreviations.
2. Making the timezone function handle things more verbosely like `America/Chicago`.
3. ?
Would you be willing to look into this further?

Comment (jz448) [reply crates/nu-cli/src/commands/date/to_timezone.rs:32]: @fdncred If we use disambiguous time zone descriptions like `America/Chicago` instead of abbreviations, then it would be trivial to use the `chrono-tz` crate to parse the time zone. This means users would need to use `date to-timezone Europe/London` instead of `date to-timezone GMT`. Although it's longer to type, it removes the ambiguity and the need to remember the time zone abbreviations. Should I go for this option?

If we still want to support time zone abbreviations, one solution I can think of is to provide users a list of ambiguous time zones' full name and ask users to choose from it, but I don't think `Nushell` currently supports such interactive session.

Comment (fdncred) [reply crates/nu-cli/src/commands/date/to_timezone.rs:32]: @jz488 yes, i think we should move ahead with a more verbose time zone functionality in order to disambiguate and still keep this feature in the PR that you've worked so hard on. we want to be as inclusive as possible, globally. so, i think the best way to do this is to use something the `chrono-tz` already supports. i looked into that crate a bit and it looks like it supports every time zone on the planet. i'm just not sure how users figure out what time zone to use. maybe there's a functionality that we do something like list the time zones in north america, austrailia, new zealand, south america, etc so people know what timezone names are valid. if listing isn't an option then i guess we just rely on them to know what to use somehow. thoughts?

Comment (jz448) [reply crates/nu-cli/src/commands/date/to_timezone.rs:32]: @fdncred Sounds like a good idea! I had another look at the `chrono-tz` crate and it's definitely viable to do this. We can add another command `date list-timezones`, which prints a table of all the time zones supported. The table has two columns: `timezone` and `region`. The value of `region` is None if the timezone doesn't match to a region.  This way, if users want only time zones related to a particular region (e.g., `Europe`), they can run something like `date list-timezones | where region == Europe`. Thoughts?

Comment (fdncred) [reply crates/nu-cli/src/commands/date/to_timezone.rs:32]: Excellent!  Make it so, number one. :)

**Comment by jz448** [comment]
@fdncred As we discussed, in this commit I have added the `date list-timezone` command:

```
> date list-timezone | where timezone =~ US | first 5
───┬─────────────────
 # │    timezone     
───┼─────────────────
 0 │ US/Alaska       
 1 │ US/Aleutian     
 2 │ US/Arizona      
 3 │ US/Central      
 4 │ US/East-Indiana 
───┴─────────────────
```

Now the `date to-timezone <timezone>` can take a timezone in the table, or `local`, or an offset (e.g., `+0600`):

```
> date now | date to-timezone US/Hawaii
2020-12-11 21:39:54.681663393 -10:00
```

**Comment by fdncred** [comment]
+1 for this change and your willingness to work with us. thanks.
